File: androidApp/src/main/java/com/rrbrambley/flashcards/home/ui/HomeViewModel.kt
package com.rrbrambley.flashcards.home.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.rrbrambley.flashcards.R
import com.rrbrambley.flashcards.auth.FeatureFlagRepository
import com.rrbrambley.flashcards.auth.FeatureFlags
import com.rrbrambley.flashcards.core.StringProvider
import com.rrbrambley.flashcards.shared.api.FlashcardApiClient
import com.rrbrambley.flashcards.shared.domain.HomeRepository
import com.rrbrambley.flashcards.shared.domain.PracticeSessionRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.ZoneId
import javax.inject.Inject

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val homeRepository: HomeRepository,
    private val practiceSessionRepository: PracticeSessionRepository,
    private val apiClient: FlashcardApiClient,
    private val stringProvider: StringProvider,
    private val featureFlagRepository: FeatureFlagRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow<HomeUiState>(HomeUiState.Loading)
    val uiState = _uiState.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    // Overall practice streak (FLA-106); null until loaded / when there's no active streak.
    private val _streak = MutableStateFlow<Int?>(null)
    val streak: StateFlow<Int?> = _streak.asStateFlow()

    // Whether the streak-calendar feature is enabled for this user (FLA-174); false until resolved.
    // The Android streak calendar (FLA-172) will render behind this.
    private val _streakCalendarEnabled = MutableStateFlow(false)
    val streakCalendarEnabled: StateFlow<Boolean> = _streakCalendarEnabled.asStateFlow()

    // One-shot user-facing messages (e.g. a failed background refresh), surfaced as a snackbar.
    private val _userMessages = MutableSharedFlow<String>(extraBufferCapacity = 1)
    val userMessages: SharedFlow<String> = _userMessages.asSharedFlow()

    private var observeJob: Job? = null

    init {
        observeHome()
        loadStreak()
        loadFlags()
    }

    /** Best-effort overall-streak fetch; a failure (or no streak) simply leaves the badge hidden. */
    private fun loadStreak() {
        viewModelScope.launch {
            _streak.value = runCatching {
                apiClient.getStreaks(ZoneId.systemDefault().id).overall.current
            }.getOrNull()
        }
    }

    /** Resolves the feature flags this screen cares about (FLA-174); off when unknown or on failure. */
    private fun loadFlags() {
        viewModelScope.launch {
            _streakCalendarEnabled.value = featureFlagRepository.isEnabled(FeatureFlags.STREAK_CALENDAR)
        }
    }

    /**
     * (Re)subscribes to the offline-first home feed: the repository emits cached data first, then
     * the backend feed. If the backend fetch fails after we've already shown cached data, we keep
     * it on screen and warn via a snackbar; only a failure with nothing cached is a hard error.
     */
    private fun observeHome() {
        observeJob?.cancel()
        observeJob = viewModelScope.launch {
            homeRepository.observeHomeData()
                .catch {
                    _isRefreshing.value = false
                    if (_uiState.value is HomeUiState.ShowHome) {
                        _userMessages.tryEmit(stringProvider.getString(R.string.home_refresh_error))
                    } else {
                        _uiState.update { HomeUiState.LoadingFailed }
                    }
                }
                .collect { homeData ->
                    _uiState.update { HomeUiState.ShowHome(homeData) }
                    _isRefreshing.value = false
                }
        }
    }

    /**
     * Discards an in-progress session (the home "×" action, FLA-205). Offline-first: the shared repo
     * tombstones the session locally (so its card drops immediately) and flushes the backend DELETE on
     * reconnect. On an unexpected failure we surface a snackbar; the observe flow refreshes the feed.
     */
    fun removeSession(sessionId: Long) {
        viewModelScope.launch {
            runCatching { practiceSessionRepository.deleteSession(sessionId) }
                .onFailure { _userMessages.tryEmit(stringProvider.getString(R.string.home_remove_session_error)) }
        }
    }

    /** Pull-to-refresh: keep the current feed visible while we re-fetch. */
    fun refresh() {
        _isRefreshing.value = true
        observeHome()
        loadStreak()
        loadFlags()
    }

    /** Retry after a load failure: show the loading state, then re-subscribe. */
    fun retry() {
        _uiState.update { HomeUiState.Loading }
        observeHome()
    }
}

File: iosApp/Flashcards/Home/HomeViewModel.swift
import Shared
import SwiftUI

/// Renders the home feed from the shared `HomeRepository` (backend `GET /home`, with an offline
/// fallback derived from cached sessions + static items).
@MainActor
final class HomeViewModel: ObservableObject {
    @Published private(set) var state: LoadState<[HomeData]> = .loading
    /// True when the latest background refresh failed but cached data is still shown — the view
    /// surfaces an unobtrusive banner (parity with Android's "couldn't refresh" snackbar).
    @Published private(set) var refreshFailed = false
    /// Overall practice streak (FLA-106); nil until loaded / when there's no active streak.
    @Published private(set) var streak: Int?

    private let repository: HomeRepository
    private let apiClient: FlashcardApiClient

    init(repository: HomeRepository, apiClient: FlashcardApiClient) {
        self.repository = repository
        self.apiClient = apiClient
    }

    /// Best-effort overall-streak fetch; a failure (or no streak) just leaves the badge hidden.
    func loadStreak() async {
        if let result = try? await apiClient.getStreaks(tz: TimeZone.current.identifier) {
            streak = Int(result.overall.current)
        }
    }

    /// The offline-first feed emits the local (cached) data first, then the backend feed. The stream
    /// stays alive through a backend outage (FLA-210): a failed refresh arrives as
    /// `HomeFeed.refreshFailed` (→ banner) with the local feed intact, so local changes keep flowing.
    /// `catch` only fires on a truly fatal stream error.
    func observe() async {
        do {
            for try await feed in asyncThrowingStream(BridgingKt.homeAdapter(repository)) {
                guard let feed = feed as? HomeFeed else { continue }
                state = .loaded(feed.cards)
                refreshFailed = feed.refreshFailed
            }
        } catch {
            markRefreshFailed()
        }
    }

    /// Pull-to-refresh: re-subscribe and await the backend feed (the stream emits cached data first,
    /// then the server feed), so the spinner reflects a real refresh rather than just the cache.
    func refresh() async {
        do {
            var emissions = 0
            for try await feed in asyncThrowingStream(BridgingKt.homeAdapter(repository)) {
                guard let feed = feed as? HomeFeed else { continue }
                state = .loaded(feed.cards)
                refreshFailed = feed.refreshFailed
                emissions += 1
                // First emission is the cache, second is the backend feed — stop once refreshed.
                if emissions >= 2 { return }
            }
        } catch {
            markRefreshFailed()
        }
    }

    private func markRefreshFailed() {
        if case .loaded = state {
            refreshFailed = true
        } else {
            state = .failed("Couldn't load your home feed. Check your connection.")
        }
    }
}
